### PR TITLE
Add incremental QOS config tests

### PR DESCRIPTION
**Pull request by isabelmsft**
### Description of PR

This PR will remain in draft mode until YANG model validation errors are resolved. Presently, the test case only runs successfully when the following YANG models are manually removed:
sonic-bgp-neighbor.yang
sonic-device_metadata.yang
sonic-feature.yang
sonic-flex_counter.yang
sonic-scheduler.yang
sonic-queue.yang
sonic-buffer-port-egress-profile-list.yang
sonic-buffer-port-ingress-profile-list.yang

Summary: Testcases of incremental qos configs for generic updater apply-patch
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
[End to End test support for Generic Updater apply-patch](https://github.com/Azure/SONiC/blob/master/doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md)

### CODE REVIEW

File: tests/generic_config_updater/test_incremental_qos.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    """
+    logger.info("Backing up config_db.json")
+    duthost.shell(BACKUP_CONFIG_DB_CMD)
+    verifyOrchagentRunningOrAssert(duthost)
+
+    yield
+
+    verifyOrchagentRunningOrAssert(duthost)
+    logger.info("Restoring config_db.json")
+    duthost.shell(RESTORE_CONFIG_DB_CMD)
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: added config reload command

File: tests/generic_config_updater/test_incremental_qos.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+
+def verifyOrchagentRunningOrAssert(duthost):
+    """
+    Verifies that orchagent is running, asserts otherwise
+
+    Args: 
+        duthost: Device Under Test (DUT)
+    """
+    result = duthost.shell(argv=["pgrep", "orchagent"])
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: changed

File: tests/generic_config_updater/test_incremental_qos.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+    yield
+
+    verifyOrchagentRunningOrAssert(duthost)
+    logger.info("Restoring config_db.json")
+    duthost.shell(RESTORE_CONFIG_DB_CMD)
+    duthost.shell(DELETE_BACKUP_CONFIG_DB_CMD)
+
+    for temp_file in INCREMENTAL_QOS_TEST_FILE_LIST:
+        duthost.shell('rm -rf {}'.format(temp_file))
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: oh haha, yes thank you. I left the r there by habit accidentally

File: tests/generic_config_updater/test_incremental_qos.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    Args:
+        duthost: DUT host object
+        key: BUFFER_POOL table key to configure
+        field: BUFFER_POOL table field to configure
+        value: BUFFER_POOL table value to be set
+    """
+    logger.info("Setting configdb field: {} to value: {}".format(field, value))
+   
+    if value:
+        cmd = "redis-cli -n 4 hset \"BUFFER_POOL|{}\" \"{}\" \"{}\" ".format(key, field, value)
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: Changed to use `sonic-db-cli`

File: tests/generic_config_updater/test_incremental_qos.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+    
+    Args:
+        duthost: DUT host object
+        patch_path: path of JSON patch to be applied
+    """
+    cmds = 'config apply-patch {}'.format(patch_path)
+
+    logger.info("Commands: {}".format(cmds))
+    output = duthost.shell(cmds)
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: Changed to reuse Jingwen's helper function, which checks `rc`

File: tests/generic_config_updater/test_incremental_qos.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    Args:
+        duthost: DUT host object
+    """
+    logger.info("Backing up config_db.json")
+    duthost.shell(BACKUP_CONFIG_DB_CMD)
+    verifyOrchagentRunningOrAssert(duthost)
+
+    yield
+
+    verifyOrchagentRunningOrAssert(duthost)
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py:57]: Yes, there is a relevant script we may use that needs to be validated before using in our test. I will follow up with Neetha.

File: tests/generic_config_updater/test_incremental_qos.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    ensure_patch_application(DUT_DEL_POOL_JSON_FILE)
+
+
+def test_set_nonexistent_pool_size_egress(duthost, ensure_dut_readiness):
+    prepare_configdb_field("egress_lossy_pool", "size", "")
+
+    set_pool_size_json = [
+        {
+            "op": "replace",
+            "path": "/BUFFER_POOL/egress_lossy_pool/size",
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: I changed to parametrize one test function, which iterates over configdb_field, existing vs nonexistent field before test, and operation

File: tests/generic_config_updater/test_incremental_qos.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+from tests.common.helpers.assertions import pytest_assert
+
+pytestmark = [
+    pytest.mark.topology('t0'),
+    pytest.mark.asic("mellanox")
+]
+
+logger = logging.getLogger(__name__)
+
+BACKUP_CONFIG_DB_CMD = "sudo cp /etc/sonic/config_db.json /etc/sonic/config_db.json.incremental_qos_orig"
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: Changed to use temp file

File: tests/generic_config_updater/test_incremental_qos.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+   
+    if value:
+        cmd = "redis-cli -n 4 hset \"BUFFER_POOL|{}\" \"{}\" \"{}\" ".format(key, field, value)
+    else:
+        cmd = "redis-cli -n 4 del \"BUFFER_POOL|{}\" \"{}\" ".format(key, field)
+   
+    verifyOrchagentRunningOrAssert(duthost)
+
+
+def ensure_patch_application(duthost, patch_path):
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: Changed to reuse Jingwen's common function

File: tests/generic_config_updater/test_incremental_qos.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+logger = logging.getLogger(__name__)
+
+BACKUP_CONFIG_DB_CMD = "sudo cp /etc/sonic/config_db.json /etc/sonic/config_db.json.incremental_qos_orig"
+RESTORE_CONFIG_DB_CMD = "sudo cp /etc/sonic/config_db.json.incremental_qos_orig /etc/sonic/config_db.json"
+DELETE_BACKUP_CONFIG_DB_CMD = "sudo rm /etc/sonic/config_db.json.incremental_qos_orig"
+INCREMENTAL_QOS_TEST_FILE_LIST = []
+
+
+def verifyOrchagentRunningOrAssert(duthost):
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: Moved to common file dut_utils.py

File: tests/generic_config_updater/test_incremental_qos.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+
+    for temp_file in INCREMENTAL_QOS_TEST_FILE_LIST:
+        duthost.shell('rm -rf {}'.format(temp_file))
+
+    logger.info("TEARDOWN COMPLETED")
+
+
+def prepare_configdb_field(duthost, key, field, value):
+    """
+    Prepares config db by setting BUFFER_POOL key and field to specified value. If value is empty, delete the current entry.
```
[the remark this answers is not in the document]
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: changed to specify empty string

[the remark this answers is not in the document]
Comment (qiluo-msft) [reply tests/generic_config_updater/test_incremental_qos.py]: It is different from the code. If value is None, you also delete.

File: tests/common/helpers/dut_utils.py [hunk 12]
```diff
@@ -260,3 +260,22 @@ def verify_features_state(duthost):
...
+    """
+    Verifies that orchagent is running, asserts otherwise
+
+    Args: 
+        duthost: Device Under Test (DUT)
+    """
+   
+    def _orchagent_running(): 
+        cmds = 'docker exec swss supervisorctl status orchagent'
+        output = duthost.shell(cmds)['stdout']
```
Comment (qiluo-msft) [inline tests/common/helpers/dut_utils.py:275]: >shell [](http://example.com/codeflow?start=25&length=5)

Could you check return code before checking stdout?

**Review by qiluo-msft — APPROVED** [review]

**Review by wangxin — APPROVED** [review]

**Comment by wangxin** [comment]
@isabelmsft In PR description, this change is related with YANG model. Is it OK to merge this PR now? Will it fail on 202012 or master branch?

**Comment by isabelmsft** [comment]
> @isabelmsft In PR description, this change is related with YANG model. Is it OK to merge this PR now? Will it fail on 202012 or master branch?

Yes, it's okay to merge now since Mohamed added the option to ignore problematic YANG models. \

Neetha has now shared the formulas for calculating proper field values (as opposed to the arbitrary values used in this PR). I have finished updating the test with proper field values. After merging this, I will open a new PR which uses the formulas so that those changes are easier to distinguish and review.

**Pull request merged**